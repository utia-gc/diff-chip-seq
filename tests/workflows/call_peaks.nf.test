nextflow_workflow {

    name "Test Workflow CALL_PEAKS"
    script "workflows/call_peaks.nf"
    workflow "CALL_PEAKS"

    test("CALL_PEAKS calls peaks for paired ChIP and Input files -- macs3 separate replicates.") {
        // instantiate helper objects
        def readsChIPPERep1Lane1  = new ReadsChIPPERep1Lane1()
        def readsInputPERep1Lane1 = new ReadsInputPERep1Lane1()
        def readsChIPPERep2Lane1  = new ReadsChIPPERep2Lane1()
        def readsInputPERep2Lane1 = new ReadsInputPERep2Lane1()
        def refs = new RefsEnsembl()


        when {
            params {
                publishDirReports = "${outputDir}/reports"
                publishDirData    = "${outputDir}/data"
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [
                            sampleName:   "${readsChIPPERep1Lane1.getSampleName()}",
                            lane:         "${readsChIPPERep1Lane1.getLane()}",
                            replicate:    "${readsChIPPERep1Lane1.getReplicate()}",
                            target:       "${readsChIPPERep1Lane1.getTarget()}",
                            mode:         "${readsChIPPERep1Lane1.getMode()}",
                            controlType:  "${readsChIPPERep1Lane1.getControlType()}",
                            readType:     "${readsChIPPERep1Lane1.getReadType()}",
                            rgFields:     "${readsChIPPERep1Lane1.getRGFields()}",
                        ],
                        file("${projectDir}/${readsChIPPERep1Lane1.getBam()}"),
                        file("${projectDir}/${readsChIPPERep1Lane1.getBamIndex()}")
                    ],
                    [
                        [
                            sampleName:   "${readsChIPPERep2Lane1.getSampleName()}",
                            lane:         "${readsChIPPERep2Lane1.getLane()}",
                            replicate:    "${readsChIPPERep2Lane1.getReplicate()}",
                            target:       "${readsChIPPERep2Lane1.getTarget()}",
                            mode:         "${readsChIPPERep2Lane1.getMode()}",
                            controlType:  "${readsChIPPERep2Lane1.getControlType()}",
                            readType:     "${readsChIPPERep2Lane1.getReadType()}",
                            rgFields:     "${readsChIPPERep2Lane1.getRGFields()}",
                        ],
                        file("${projectDir}/${readsChIPPERep2Lane1.getBam()}"),
                        file("${projectDir}/${readsChIPPERep2Lane1.getBamIndex()}")
                    ],
                    [
                        [
                            sampleName:   "${readsInputPERep1Lane1.getSampleName()}",
                            lane:         "${readsInputPERep1Lane1.getLane()}",
                            replicate:    "${readsInputPERep1Lane1.getReplicate()}",
                            target:       "${readsInputPERep1Lane1.getTarget()}",
                            mode:         "${readsInputPERep1Lane1.getMode()}",
                            controlType:  "${readsInputPERep1Lane1.getControlType()}",
                            readType:     "${readsInputPERep1Lane1.getReadType()}",
                            rgFields:     "${readsInputPERep1Lane1.getRGFields()}",
                        ],
                        file("${projectDir}/${readsInputPERep1Lane1.getBam()}"),
                        file("${projectDir}/${readsInputPERep1Lane1.getBamIndex()}")
                    ],
                    [
                        [
                            sampleName:   "${readsInputPERep2Lane1.getSampleName()}",
                            lane:         "${readsInputPERep2Lane1.getLane()}",
                            replicate:    "${readsInputPERep2Lane1.getReplicate()}",
                            target:       "${readsInputPERep2Lane1.getTarget()}",
                            mode:         "${readsInputPERep2Lane1.getMode()}",
                            controlType:  "${readsInputPERep2Lane1.getControlType()}",
                            readType:     "${readsInputPERep2Lane1.getReadType()}",
                            rgFields:     "${readsInputPERep2Lane1.getRGFields()}",
                        ],
                        file("${projectDir}/${readsInputPERep2Lane1.getBam()}"),
                        file("${projectDir}/${readsInputPERep2Lane1.getBamIndex()}")
                    ],
                )
                input[1] = file("${projectDir}/${refs.getGenome()}")
                """
            }
        }

        then {
            // test workflow status
            assert workflow.success
            assert workflow.trace.succeeded().size() == 3

            // test peak calling logs for each peak set
            with(workflow.out.callPeaksLog) {
                // test rep1 files
                assert size() == 2
                with(get(0)) {
                    assert size() == 2
                    // test peak calling log
                    String sampleName = [
                        readsChIPPERep1Lane1.getSampleName(),
                        "${readsChIPPERep1Lane1.getTarget()}-vs-${readsChIPPERep1Lane1.getControlType()}",
                        "rep${readsChIPPERep1Lane1.getReplicate()}",
                    ].join('_')
                    assert get(1) ==~ /^.*\/${sampleName}_peaks\.xls/
                }
                // test rep2 files
                with(get(1)) {
                    assert size() == 2
                    // test peak calling log
                    String sampleName = [
                        readsChIPPERep2Lane1.getSampleName(),
                        "${readsChIPPERep2Lane1.getTarget()}-vs-${readsChIPPERep2Lane1.getControlType()}",
                        "rep${readsChIPPERep2Lane1.getReplicate()}",
                    ].join('_')
                    assert get(1) ==~ /^.*\/${sampleName}_peaks\.xls/
                }
            }

            // test reproducibility
            assert snapshot(workflow.out).match()
        }

    }

}
