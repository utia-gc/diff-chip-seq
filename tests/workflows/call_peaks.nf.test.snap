{
    "CALL_PEAKS calls peaks for paired ChIP and Input files -- macs3.": {
        "content": [
            {
                "0": [
                    [
                        {
                            "sampleName": "wt_ip1",
                            "replicate": "1",
                            "target": "FLAG-IP1",
                            "controlType": "input",
                            "readType": "paired",
                            "rgFields": "[ID:wt_ip1.001, SM:wt_ip1, LB:wt_ip1, PL:ILLUMINA]"
                        },
                        "wt_ip1_FLAG-IP1-vs-input_rep1_peaks.xls:md5,f7e75b4108eddc22a190c884aa97f402"
                    ],
                    [
                        {
                            "sampleName": "wt_ip1",
                            "replicate": "2",
                            "target": "FLAG-IP1",
                            "controlType": "input",
                            "readType": "paired",
                            "rgFields": "[ID:wt_ip1.001, SM:wt_ip1, LB:wt_ip1, PL:ILLUMINA]"
                        },
                        "wt_ip1_FLAG-IP1-vs-input_rep2_peaks.xls:md5,26999449f557f8462973464a3a362e5f"
                    ]
                ],
                "callPeaksLog": [
                    [
                        {
                            "sampleName": "wt_ip1",
                            "replicate": "1",
                            "target": "FLAG-IP1",
                            "controlType": "input",
                            "readType": "paired",
                            "rgFields": "[ID:wt_ip1.001, SM:wt_ip1, LB:wt_ip1, PL:ILLUMINA]"
                        },
                        "wt_ip1_FLAG-IP1-vs-input_rep1_peaks.xls:md5,f7e75b4108eddc22a190c884aa97f402"
                    ],
                    [
                        {
                            "sampleName": "wt_ip1",
                            "replicate": "2",
                            "target": "FLAG-IP1",
                            "controlType": "input",
                            "readType": "paired",
                            "rgFields": "[ID:wt_ip1.001, SM:wt_ip1, LB:wt_ip1, PL:ILLUMINA]"
                        },
                        "wt_ip1_FLAG-IP1-vs-input_rep2_peaks.xls:md5,26999449f557f8462973464a3a362e5f"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-09-15T18:16:26.814273425"
    }
}